frame/nomination-pools/src/lib.rs

```diff
@@ -0,0 +1,1967 @@
+//!
+//! **Relevant extrinsics:**
+//!
+//! * [`Call::unbond_other`]
+//! * [`Call::withdraw_unbonded_other`]
+//!
+//! ### Slashing
+//!
+//! This section assumes that the slash computation is executed by
+//! [`pallet_staking::StakingLedger::slash`], which passes the information to this pallet via
+//! [`sp_staking::OnStakerSlash::on_slash`].
+//!
+//! Unbonding pools need to be slashed to ensure all nominators whom where in the bonded pool
+//! while it was backing a validator that equivocated are punished. Without these measures a
+//! nominator could unbond right after a validator equivocated with no consequences.
+//!
+//! This strategy is unfair to delegators who joined after the slash, because they get slashed as
+//! well, but spares delegators who unbond. The latter is much more important for security: if a
+//! pool's validators are attacking the network, their delegators need to unbond fast! Avoiding
+//! slashes gives them an incentive to do that if validators get repeatedly slashed.
+//!
+//! To be fair to joiners, this implementation also need joining pools, which are actively staking,
+//! in addition to the unbonding pools. For maintenance simplicity these are not implemented.
+//! Related: https://github.com/paritytech/substrate/issues/10860
+//!
+//! **Relevant methods:**
+//!
+//! * [`Pallet::do_slash`]
+//!
+//! ### Limitations
+//!
+//! * Delegators cannot vote with their staked funds because they are transferred into the pools
+//!   account. In the future this can be overcome by allowing the delegators to vote with their
+//!   bonded funds via vote splitting.
+//! * Delegators cannot quickly transfer to another pool if they do no like nominations, instead
+//!   they must wait for the unbonding duration.
+//!
+//! # Runtime builder warnings
+//!
+//! * Watch out for overflow of [`RewardPoints`] and [`BalanceOf`] types. Consider things like the
+//!   chains total issuance, staking reward rate, and burn rate.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Codec;
+use frame_support::{
+	ensure,
+	pallet_prelude::{MaxEncodedLen, *},
+	storage::bounded_btree_map::BoundedBTreeMap,
+	traits::{
+		Currency, Defensive, DefensiveOption, DefensiveResult, DefensiveSaturating,
+		ExistenceRequirement, Get,
+	},
+	DefaultNoBound, RuntimeDebugNoBound,
+};
+use scale_info::TypeInfo;
+use sp_core::U256;
+use sp_runtime::traits::{AccountIdConversion, Bounded, Convert, Saturating, Zero};
+use sp_staking::{EraIndex, OnStakerSlash, StakingInterface};
+use sp_std::{collections::btree_map::BTreeMap, fmt::Debug, ops::Div, vec::Vec};
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+pub type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+pub type SubPoolsWithEra<T> = BoundedBTreeMap<EraIndex, UnbondPool<T>, TotalUnbondingPools<T>>;
+pub type RewardPoints = U256;
```

Comment:
missing docs for `RewardPoints` and `Points`. 

The other two need not be public IMO.

Reply:
I will double check, but I think they are all pub for the benchmarking crate

Reply:
SubPoolsWithEra is the only one not used in benchmarks